base/datafmt.jl

```diff
@@ -18,12 +18,13 @@ const offs_chunk_size = 5000
 countlines(f::AbstractString, eol::Char='\n') = open(io->countlines(io,eol), f)::Int
 function countlines(io::IO, eol::Char='\n')
     isascii(eol) || throw(ArgumentError("only ASCII line terminators are supported"))
+    aeol = UInt8(eol)
     a = Array(UInt8, 8192)
     nl = 0
     while !eof(io)
         nb = readbytes!(io, a)
         @simd for i=1:nb
-            @inbounds nl += a[i] == eol
+            @inbounds nl += a[i] == aeol
```

Comment:
I just tested it, and the more natural `nl += a[i] == UInt8(eol)` gives the same native code.